### PR TITLE
Tremblp/maya 114982/maya reference 00

### PR DESCRIPTION
**Pull request by ppt-adsk**

### CODE REVIEW

File: lib/mayaUsd/fileio/primUpdater.h [hunk 1]
```diff
@@ -55,10 +55,19 @@ class UsdMayaPrimUpdater
...
+        Failed,
+        Continue,
+        Prune
+    };
+
     // Copy the pushed prim from the temporary srcLayer where it has been
     // exported by push into the destination dstLayer which is in the scene.
     MAYAUSD_CORE_PUBLIC
-    virtual bool pushCopySpecs(
+    virtual PushCopySpecs pushCopySpecs(
```
Comment (ppt-adsk) [inline lib/mayaUsd/fileio/primUpdater.h:70]: pushCopySpecs() must indicate behavior beyond simple success and failure: pruning traversal is also success.

File: lib/mayaUsd/fileio/primUpdaterManager.cpp [hunk 2]
```diff
@@ -575,33 +581,47 @@ SdfPath getDstSdfPath(const Ufe::Path& ufePulledPath, const SdfPath& srcSdfPath,
...
     const SdfLayerRefPtr&            srcLayer,
-    const SdfPath&                   primSpecPath,
+    const SdfPath&                   srcPath,
+    const SdfLayerRefPtr&            dstLayer,
+    const SdfPath&                   dstPath,
     const UsdMayaPrimUpdaterContext& context)
 {
-    // Get the primSpec from the src layer.
-    auto primSpec = srcLayer->GetPrimAtPath(primSpecPath);
+    // The root of the pulled hierarchy is crucial for determining push
```
Comment (ppt-adsk) [inline lib/mayaUsd/fileio/primUpdaterManager.cpp:591]: Having a stable updater that performs both the pull and the push is crucial to success.  Ensure we get the same updater by using the original pulled path to create the updater for both pull and push.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel [hunk 3]
```diff
@@ -0,0 +1,640 @@
+// Verbatim copy of fileOptions.mel:pushOptionsUITemplate(), which is not
```
Comment (ppt-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:1]: Main file in the review.  Unfortunately mostly copy-pasted and adapted from Maya code, which was not designed to be extensible to use cases like Maya reference prim creation.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import maya.api.OpenMaya as om
+from pxr import Sdf
+import mayaUsd
+import ufe
+
+# Adapted from testMayaUsdSchemasMayaReference.py.
+def createMayaReferencePrim(ufePathStr, mayaReferencePath, mayaNamespace):
```
Comment (ppt-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py:7]: Once all the UI dialog is done, perform the Maya reference prim creation.

File: lib/mayaUsd/ufe/UsdContextOps.cpp [hunk 5]
```diff
@@ -375,12 +377,12 @@ global proc string ClearAllUSDReferencesConfirm()
 ClearAllUSDReferencesConfirm();
 )";
 
-class AddReferenceUndoableCommand : public Ufe::UndoableCommand
+class AddUsdReferenceUndoableCommand : public Ufe::UndoableCommand
```
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/UsdContextOps.cpp:380]: Renamed for clarity: existing code creates a USD reference, not a Maya reference prim.

File: lib/usd/translators/mayaReferenceUpdater.cpp [hunk 6]
```diff
@@ -45,7 +45,11 @@ PXRUSDMAYA_REGISTER_UPDATER(
     reference,
     PxrUsdTranslators_MayaReferenceUpdater,
     (UsdMayaPrimUpdater::Supports::Push | UsdMayaPrimUpdater::Supports::Clear
-     | UsdMayaPrimUpdater::Supports::AutoPull));
+     // Disable AutoPull pending more granular support for it: MayaReference
```
Comment (ppt-adsk) [inline lib/usd/translators/mayaReferenceUpdater.cpp:48]: Change is out of scope for this pull request: auto pull is not based on type, it's based on an instance of a type, i.e. because we need to ask the instance whether it can auto pull or not (and an instance without a file path should not be auto pulled).  We will need a virtual function to customize this.

File: modules/mayaUSD.mod.template [hunk 7]
```diff
@@ -13,7 +13,6 @@ PXR_MTLX_STDLIB_SEARCH_PATHS+=${USD_INSTALL_LOCATION}/libraries
 icons: lib/icons
 plug-ins: 
 presets: 
-scripts:
```
Comment (ppt-adsk) [inline modules/mayaUSD.mod.template:16]: scripts folder is no longer empty.

File: lib/mayaUsd/fileio/primUpdater.h [hunk 8]
```diff
@@ -55,10 +55,19 @@ class UsdMayaPrimUpdater
         All = Push | Pull | Clear
     };
 
+    // pushCopySpecs result code.  Prune means success, but no further
+    // traversal should take place.
+    enum class PushCopySpecs
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primUpdater.h:60]: It feels weird to have a enum have the same name as a function. Maybe PushCopyResult would have been a better name? In any case, this is not a widely used type, so it is not very important.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    '''Create a Maya reference prim parented to the argument path.'''
+
+    # Append reference name to parent Sdf path.
+    ufePath = ufe.PathString.path(ufePathStr)
+    parentPath = ufePath.segments[1]
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py:12]: Could you explain the hard-coded [1] here? If I understand it, the code assumes the second segment is a USD path and adds '/MayaReference'  to it? I found the comment is not entirely clear. I think the comment would be more helpful if it said something like:

Extract the USD path segment from the UFE path and adds the MayaReference under it

Or something similar.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    primPath = Sdf.AssetPath(str(parentPath)+'/MayaReference')
+
+    stage = mayaUsd.ufe.getStage(ufePathStr)
+    prim = stage.DefinePrim(primPath.path, 'MayaReference')
+    if not prim.IsValid():
+        om.MGlobal.displayError("Could not create MayaReference prim under %s" % ufePathStr)
+        return
+
+    mayaReferenceAttr = prim.CreateAttribute('mayaReference', Sdf.ValueTypeNames.Asset)
+    mayaReferenceAttr.Set(mayaReferencePath)
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py:22]: We got a MayaReference path, a MayaReference custom prim type and mayaReference attribute. It's a bit confusing :) I suppose it cannot be avoided.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel (hunk 3 above)
Comment (pierrebai-adsk) [reply lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:1]: I'm honestly not very knowledgeable about MEL and always struggle to write it, so I'm not going to pretend to review it. This being mostly UI code, I'll approve it blindly.

**Review by pierrebai-adsk — CHANGES_REQUESTED** [review]
Just waiting for your feedback on my suggestions.

File: lib/mayaUsd/fileio/primUpdaterManager.cpp [hunk 11]
```diff
@@ -575,33 +581,47 @@ SdfPath getDstSdfPath(const Ufe::Path& ufePulledPath, const SdfPath& srcSdfPath,
...
+    // behavior.  When pulling, we may have created a Maya pull hierarchy root
+    // node whose type does not map to the same prim updater as the original
+    // USD prim, i.e. multiple USD prim types can map to the same pulled Maya
+    // node type (e.g. transform, which is the fallback Maya node type for many
+    // USD prim types).  Therefore, if we're at the root of the src hierarchy,
+    // use the prim at the pulled path to create the prim updater.
+    const bool usePulledPrim = (srcPath.GetPathElementCount() == 1);
```
[the remark this answers is not in the document]
Comment (ppt-adsk) [reply lib/mayaUsd/fileio/primUpdaterManager.cpp:598]: Correct.  I'll adjust the comment to mention the temporary layer.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py (hunk 10 above)
Comment (ppt-adsk) [reply lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py:22]: Those names were chosen by Animal Logic, I won't change them.
https://github.com/Autodesk/maya-usd/blob/600c31abcd68ef3fe0047b28fb5f33f6edd0872f/lib/usd/schemas/schema.usda.in#L22

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py (hunk 9 above)
Comment (ppt-adsk) [reply lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py:12]: O.K.

**Review by pierrebai-adsk — APPROVED** [review]
